realm/realm-library/src/main/java/io/realm/DynamicRealmObject.java

```diff
@@ -115,6 +115,7 @@ public DynamicRealmObject(RealmModel obj) {
      */
     public boolean getBoolean(String fieldName) {
         long columnIndex = proxyState.getRow$realm().getColumnIndex(fieldName);
+        checkFieldType(fieldName, columnIndex, RealmFieldType.BOOLEAN);
```

Comment:
Isn't the type checked in C++ on the JNI side when calling getX()? Otherwise it should as that would be more performant.